cpp/src/io/json/json_tree.cu

```diff
@@ -532,14 +583,23 @@ std::pair<rmm::device_uvector<size_type>, rmm::device_uvector<size_type>> hash_n
   auto const d_equal = [node_level      = node_levels.begin(),
                         node_type       = node_type.begin(),
                         parent_node_ids = parent_node_ids.begin(),
+                        is_array_of_arrays,
+                        row_array_children_level,
+                        list_indices = list_indices.begin(),
                         d_hashed_cache] __device__(auto node_id1, auto node_id2) {
     if (node_id1 == node_id2) return true;
     if (d_hashed_cache(node_id1) != d_hashed_cache(node_id2)) return false;
-    auto const is_equal_level = [node_level, node_type](auto node_id1, auto node_id2) {
-      if (node_id1 == node_id2) return true;
-      return node_level[node_id1] == node_level[node_id2] and
-             node_type[node_id1] == node_type[node_id2];
-    };
+    auto const is_equal_level =
+      [node_level, node_type, is_array_of_arrays, row_array_children_level, list_indices](
+        auto node_id1, auto node_id2) {
+        if (node_id1 == node_id2) return true;
+        bool const is_level2_equal =
+          is_array_of_arrays and node_level[node_id1] == row_array_children_level
+            ? list_indices[node_id1] == list_indices[node_id2]
+            : true;
```

Comment:
This is simpler in theory, not sure if it's actually clearer
```suggestion
          !(is_array_of_arrays and node_level[node_id1] == row_array_children_level) or
            list_indices[node_id1] == list_indices[node_id2];
```
I'm surprised that `is_level2_equal` can be true only when `node_level[node_id1] != row_array_children_level`. Maybe a comment would be helpful here.

Reply:
Here, ternary operator condition is kept because `list_indices` will be empty if `is_array_of_arrays==false`.
if `is_array_of_arrays==false`, then this condition should not affect overall condition, so it returns `true`.
The name `is_level2_equal` may be misleading. How about `is_top_row_same_column` or `has_same_array_index_at_row_array` ?

or is the following code clearer? (I would prefer to avoid this due to branch divergence).
```
if(is_array_of_arrays) {
return return node_level[node_id1] == node_level[node_id2] and
               node_type[node_id1] == node_type[node_id2] and 
               (node_level[node_id1] != row_array_children_level or list_indices[node_id1] == list_indices[node_id2]);
else
return node_level[node_id1] == node_level[node_id2] and
               node_type[node_id1] == node_type[node_id2];
```

Reply:
> Here, ternary operator condition is kept because list_indices will be empty if is_array_of_arrays==false.

FWIW, you can still you my version because of [short-circuiting](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#conditional-logical-and-operator-) that && and || do.

Reply:
Went over this again. The logic is convoluted enough that it would be best for the code to reflect your internal logic. Well named intermediate conditions and/or comments would also help.

Reply:
updated to lambda as discussed.